dbms/src/Storages/DeltaMerge/Segment.cpp

```diff
@@ -2224,5 +2270,111 @@ bool Segment::placeDelete(const DMContext & dm_context,
     return fully_indexed;
 }
 
+BitmapFilterPtr Segment::buildBitmapFilter(const DMContext & dm_context,
+                                           const SegmentSnapshotPtr & segment_snap,
+                                           const RowKeyRanges & read_ranges,
+                                           const RSOperatorPtr & filter,
+                                           UInt64 max_version,
+                                           size_t expected_block_size)
+{
+    Stopwatch sw_total;
+    static ColumnDefines columns_to_read{
+        getExtraHandleColumnDefine(is_common_handle),
+    };
+    auto stream = getInputStreamModeNormal(
+        dm_context,
+        columns_to_read,
+        segment_snap,
+        read_ranges,
+        filter,
+        max_version,
+        expected_block_size,
+        /*need_row_id*/ true);
+    auto total_rows = segment_snap->delta->getRows() + segment_snap->stable->getDMFilesRows();
+    auto bitmap_filter = std::make_shared<BitmapFilter>(total_rows, segment_snap);
+    for (;;)
+    {
+        FilterPtr f = nullptr;
+        auto blk = stream->read(f, true);
+        if (likely(blk))
+        {
+            bitmap_filter->set(blk.segmentRowIdCol(), f);
+        }
+        else
+        {
+            break;
+        }
+    }
+    bitmap_filter->runOptimize();
+    LOG_DEBUG(log, "buildBitmapFilter total_rows={} cost={}ms", total_rows, sw_total.elapsedMilliseconds());
+    return bitmap_filter;
+}
+
+BlockInputStreamPtr Segment::getBitmapFilterInputStream(BitmapFilterPtr && bitmap_filter,
+                                                        const DMContext & dm_context,
+                                                        const ColumnDefines & columns_to_read,
+                                                        const RowKeyRanges & data_ranges,
+                                                        const RSOperatorPtr & filter,
+                                                        UInt64 max_version,
+                                                        size_t expected_block_size)
+{
+    auto & segment_snap = bitmap_filter->snapshot();
+    auto enable_handle_clean_read = !hasColumn(columns_to_read, EXTRA_HANDLE_COLUMN_ID);
+    constexpr auto is_fast_scan = true;
+    auto enable_del_clean_read = !hasColumn(columns_to_read, TAG_COLUMN_ID);
```

Comment:
I think `enable_del_clean_read` is always false. TiDB Planner is not aware of the tag column, so columns_to_read never contain tag column.